rules/noUnnecessaryElseRule.ts

```diff
@@ -1,22 +1,54 @@
-import * as ts from 'typescript';
+// tslint
 import * as Lint from 'tslint';
-import * as utils from 'tsutils';
-import { AbstractIfStatementWalker } from '../src/walker';
+
+// tsrules.tslint-consistent-codestyle.src
 import { isElseIf } from '../src/utils';
 
+// tsutils
+import * as utils from 'tsutils';
+
+// typescript
+import * as ts from 'typescript';
+
 const FAIL_MESSAGE = `unnecessary else`;
+const FAIL_MESSAGE_BLOCK = `unnecessary else block`;
 
 export class Rule extends Lint.Rules.AbstractRule {
-    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, undefined));
-    }
+  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+    return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, undefined));
+  }
 }
 
-class IfWalker extends AbstractIfStatementWalker<void> {
-    protected _checkIfStatement(node: ts.IfStatement) {
-        if (node.elseStatement !== undefined &&
-            !isElseIf(node) &&
-            utils.endsControlFlow(node.thenStatement))
-            this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE);
+class IfWalker extends Lint.AbstractWalker<void> {
+  public walk(sourceFile: ts.SourceFile) {
+    const cb = (node: ts.Node): void => {
+      if (node.kind === ts.SyntaxKind.IfStatement && !this._checkIfStatement(<ts.IfStatement>node))
+        return;
+      return ts.forEachChild(node, cb);
+    };
+
+    return ts.forEachChild(sourceFile, cb);
+  }
+
+  /** Return true to traverse children, false otherwise */
+  private _checkIfStatement(node: ts.IfStatement): boolean {
+    // remove else
+    const elseStatement = node.elseStatement;
+    if (isElseStatement(elseStatement) && !isElseIf(node) && utils.endsControlFlow(node.thenStatement)) {
+      const fixElse = Lint.Replacement.replaceFromTo(node.thenStatement.getEnd(), elseStatement.getStart(), '');
+      this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE, fixElse);
+
+      // remove else scope if safe: we keep multi-statement blocks since the may override local variables when unwrapped
```

Comment:
technically that's not correct. consider the following (I know it doesn't make much sense):

```ts
let Foo = 1;
function canNotBeFixed() {
  if (condition) {
    return Foo;
  } else {
    class Foo {} // fixing this would cause a runtime error in the then-branch
  }
}
```

The else-block can be unwrapped if it doesn't contain any block-scoped declarations, regardless of the number of statements